storage/src/freezer/storage.rs

```diff
@@ -0,0 +1,624 @@
+    }
+}
+
+impl<const N: usize> commonware_codec::FixedSize for PrefixedArray<N> {
+    const SIZE: usize = N;
+}
+
+impl<const N: usize> commonware_codec::Write for PrefixedArray<N> {
+    fn write(&self, buf: &mut impl bytes::BufMut) {
+        buf.put_slice(&self.data);
+    }
+}
+
+impl<const N: usize> commonware_codec::Read for PrefixedArray<N> {
+    type Cfg = ();
+
+    fn read_cfg(
+        buf: &mut impl bytes::Buf,
+        _cfg: &Self::Cfg,
+    ) -> Result<Self, commonware_codec::Error> {
+        if buf.remaining() < N {
+            return Err(commonware_codec::Error::EndOfBuffer);
+        }
+        let mut data = [0u8; N];
+        buf.copy_to_slice(&mut data);
+        Ok(Self { data })
+    }
+}
+
+impl<const N: usize> Array for PrefixedArray<N> {}
+
+/// Wrapper enum for values in the unified diskmap
+#[derive(Clone, Debug)]
+enum StorageValue<K: Array, V: Codec> {
+    Value(V),
+    Key(K),
+}
+
+impl<K: Array + Codec<Cfg = ()>, V: Codec> commonware_codec::EncodeSize for StorageValue<K, V> {
+    fn encode_size(&self) -> usize {
+        match self {
+            StorageValue::Value(v) => 1 + v.encode_size(),
+            StorageValue::Key(k) => 1 + k.encode_size(),
+        }
+    }
+}
+
+impl<K: Array + Codec<Cfg = ()>, V: Codec> commonware_codec::Write for StorageValue<K, V> {
+    fn write(&self, buf: &mut impl bytes::BufMut) {
+        match self {
+            StorageValue::Value(v) => {
+                buf.put_u8(0);
+                v.write(buf);
+            }
+            StorageValue::Key(k) => {
+                buf.put_u8(1);
+                k.write(buf);
+            }
+        }
+    }
+}
+
+impl<K: Array + Codec<Cfg = ()>, V: Codec> commonware_codec::Read for StorageValue<K, V> {
+    type Cfg = V::Cfg;
+
+    fn read_cfg(
+        buf: &mut impl bytes::Buf,
+        cfg: &Self::Cfg,
+    ) -> Result<Self, commonware_codec::Error> {
+        let tag = buf.get_u8();
+        match tag {
+            0 => Ok(StorageValue::Value(V::read_cfg(buf, cfg)?)),
+            1 => Ok(StorageValue::Key(K::read_cfg(buf, &())?)),
+            _ => Err(commonware_codec::Error::InvalidEnum(tag)),
+        }
+    }
+}
+
+/// Implementation of `Freezer` storage using a single diskmap with prefixed keys.
+pub struct Freezer<E: Storage + Metrics, K: Array, V: Codec> {
+    // Single DiskMap for both key->value and index->key storage
+    // We use a large enough array to hold any prefixed key (1 byte prefix + max of key or index)
+    storage: DiskMap<E, PrefixedArray<256>, StorageValue<K, V>>,
+
+    // RMap for interval tracking (only in-memory data structure)
+    intervals: RMap,
+
+    // Metrics
+    items_tracked: Gauge,
+    gets: Counter,
+    has: Counter,
+    syncs: Counter,
+
+    // Store the actual key size for runtime validation
+    _key_size: usize,
+}
+
+impl<E: Storage + Metrics, K: Array + Codec<Cfg = ()>, V: Codec> Freezer<E, K, V> {
```

Comment:
TODO: use a flat file to store a map of index <> (key + CRC). Only iterate over this index file on startup instead of putting all indexes into the Diskmap.